libjob and job-ingest: improve submit API and submit/queue synchronization

garlick opened the pull request:
This PR solves a few problems with the `flux_job_submit()` API call and the `job-ingest` module identified in discussions and while building the `job-manager` module.

First, `flux_job_submit()` requires the user to sign jobspec before submitting it, which means the user must interact with the flux-security API.  This is extra work for little gain, since most of the time the default config file and signing method is appropriate.  In addition, flux-core can be built without flux-security, and then `flux_job_submit()` expects jobspec to be sent in the clear.  This means the user has to know how flux-core was built.

To clean this up, add a `FLUX_JOB_PRE_SIGNED` flag to indicate that jobspec is already signed when passed to `flux_job_submit()` for the uncommon case.  In the common case, this flag is not used, and the flux-security API is used internally to sign using the default configuration  When flux-core is not built with flux-security, a dumbed down wrap/unwrap interface is used internally to wrap the jobspec with the "none" method using the same encoding as flux-security.

Next, the `job-ingest` module announces new jobs with an event message, and the event payload contains only an array of jobid's.  The `job-manager` module must immediately put these into a queue ordered by priority and regulate access to them based on the userid, so it must fetch the userid and priority from the KVS.  To fix this, add userid and priority to the payload so instead of an array of integers, we send an array of objects containing id, userid, and priority.

Finally, the event created an "eventually consistent" relationship between the submit command/API and the queue which at minimum makes testing the job manager difficult, and at worst, might confuse users.  Change the announcing event to an RPC to the job manager, and add a dummy job manager module for testing.

garlick commented:
I went ahead and rebased this on current master, and also improved some commit messages.  Apologies if I did that while someone was looking at code!

SteVwonder commented:
After going through the code, it looks good to me.  I'll try and play around with this over the weekend.  See if I can throw some jobs at it with my simple python initial program.

garlick commented:
I tacked on some commits to fill in testing gaps and fix up some nits.  @SteVwonder  and @grondo:  OK to squash the incremental development and rebase on master?

grondo commented:
That is fine with me!

garlick commented:
OK, I went ahead and forced a push.

garlick commented:
Rebased again.

garlick commented:
@stevwonder brought up the KVS eventlog yesterday in our discussion, and it gradually sunk in afterward that probably the KVS eventlog _should_ be created by the `job-ingest` module after all, since it's already doing a KVS commit on the job's behalf, and the `job-manager` module shouldn't need to do anything until the next event.  So I'll add that on here.

garlick commented:
> Side note: my job eventlogs seem to have an extra line?

I think that's an artifact of `flux kvs get`.  Try `--raw` or use `flux kvs eventlog get`.

> For some reason, on instance shutdown I'm always getting a warning that job-ingest is not cleanly shutdown.

Oh, I had noticed that but had forgotten that instance shutdown is supposed to shut down modules as well.  I'll have a look.

> the job for which submit failed has an entry in the KVS and its eventlog has a submit event logged

Er, I think it should be cleaned out as though it never existed if the jobid hasn't been returned to the user yet.  Will fix that.

Thanks, good feedback!

grondo commented:
> I think that's an artifact of flux kvs get. Try --raw or use flux kvs eventlog get.

Oh yeah, duh!

SteVwonder commented on src/common/libjob/job.c:
```diff
@@ -25,26 +25,75 @@
+        if (!sec)
+            goto error;
+        if (!(J = flux_sign_wrap (sec, jobspec, strlen (jobspec), NULL, 0)))
+            goto error;
```
If this fails, can you grab/set the errnum and grab/log the errmsg from the security context?
```C
       if (!(J = flux_sign_wrap (sec, jobspec, strlen (jobspec), NULL, 0))) {
           flux_log (h, LOG_ERR, "%s: %s", __FUNCTION__, flux_security_last_error (sec));
           errno = flux_security_last_errnum (sec);
           goto error;
       }
```

garlick replied:
Good point.  Certainly the errno should pass through if it doesn't already.   We shouldn't log from a library and there is currently no way to return an error string to the user since we are not returning a future.  I suppose we could go ahead and return a future and immediately fulfill it with an error and error string in that case (also if setting up the security context fails e.g. due to bad config file).

Hmm pondering...

garlick commented:
>> For some reason, on instance shutdown I'm always getting a warning that job-ingest is not cleanly shutdown.

> Oh, I had noticed that but had forgotten that instance shutdown is supposed to shut down modules as well. I'll have a look.

Sorry, I was confused and needed to review that code in the broker.  Normally we have balanced load and remove calls in rc1 and rc3 and any loaded module not explicitly removed will trigger that error.  (rmmod sends a `<name>.shutdown` request to the module, and the default handler for this causes the reactor to exit cleanly; if this doesn't happen by the time `ctx->modhash` is destroyed, then the module thread is `pthread_cancel()`ed, and this causes the warning about not being cleanly shutdown).

grondo commented:
> Sorry, I was confused and needed to review that code in the broker. Normally we have balanced load and remove calls in rc1 and rc3 and any loaded module not explicitly removed will trigger that error. (rmmod sends a <name>.shutdown request to the module, and the default handler for this causes the reactor to exit cleanly; if this doesn't happen by the time ctx->modhash is destroyed, then the module thread is pthread_cancel()ed, and this causes the warning about not being cleanly shutdown).

Oh yeah, I had forgotten about that too and now it seems obvious. Sorry!

garlick commented:
I think I've addressed all the review comments thus far.  I did throw some flux-job cleanup just now also.

The coverage is below the threshold, though much of the missed coverage is in error handling that's hard to hit. 

One bit of missing coverage is allowing `flux_job_submit()` to internally sign jobspec.  That's somewhat tricky since (as @SteVwonder pointed out elsewhere) we don't have munge running in travis yet, and the installed config file selects munge as the default signing mechanism.  Maybe that can come later?

garlick commented:
Unless there are objections, i'm going to squash down the incremental development here.

garlick commented:
Squashed and rebased on current master.

garlick commented:
I think we agreed this could go in - could somebody please press the button?  :-)
